modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,200 @@
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.geotools.util.Utilities.stream;
+
+/*
+ * TODO: document all
+ *
+ *  - no null checks
+ *  - no subclass relationships between categories and instances
+ */
+
+class CategoryRegistry {
+
+	private final FactoryRegistry factoryRegistry;
+	private final Map<Class<?>, InstanceRegistry<?>> categories = new HashMap<>();
+
+	public CategoryRegistry(FactoryRegistry factoryRegistry) {
+		this.factoryRegistry = factoryRegistry;
+	}
+
+	public void registerCategory(final Class<?> category) {
+		// TODO: should anything happen if the category is already registered?
+		categories.put(category, new InstanceRegistry<>(factoryRegistry, category));
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T> void registerInstance(final T instance) {
+		streamCategories()
+				.filter(category -> category.isAssignableFrom(instance.getClass()))
+				// the cast is correct because the filter above only leaves categories that are supertypes of `instance`
+				.map(category -> (InstanceRegistry<? super T>) instanceRegistry(category))
+				.forEach(registry -> registry.register(instance));
+	}
+
+	// documentation hint: returns true if this the first instance of its class
+	public <T> boolean registerInstance(final T instance, final Class<T> category) {
+		return instanceRegistry(category).register(instance);
+	}
+
+	// documentation hint: returns true if an instance of the same type was previously registered
+	public <T> void deregisterInstance(final T instance) {
+		streamCategories()
+				.filter(category -> category.isAssignableFrom(instance.getClass()))
+				// the cast is correct because the filter above only leaves categories that are supertypes of `instance`
+				.map(category -> (InstanceRegistry<? super T>) instanceRegistry(category))
+				.forEach(registry -> registry.deregister(instance));
+	}
+
+	// documentation hint: returns true if an instance of the same type was previously registered
+	public <T> boolean deregisterInstance(final T instance, final Class<T> category) {
+		return instanceRegistry(category).deregister(instance);
+	}
+
+	public void deregisterInstances() {
+		categories.values().forEach(InstanceRegistry::clear);
+	}
+
+	public void deregisterInstances(Class<?> category) {
+		instanceRegistry(category).clear();
+	}
+
+	private <T> InstanceRegistry<T> instanceRegistry(final Class<T> category) {
+		@SuppressWarnings("unchecked")
+		InstanceRegistry<T> registry = (InstanceRegistry<T>) categories.get(category);
+		if (registry == null) {
+			// TODO: do something fancy like in FactoryRegistry#getServiceProvider(Class, Predicate, Hints, Hints.Key) ?
+			throw new IllegalArgumentException("The category '" + category + "' is not registered");
+		}
+		return registry;
+	}
+
+	public Stream<Class<?>> streamCategories() {
+		return categories.keySet().stream();
```

Comment:
Good point, but there is no code registering/deregistering categories after construction of `FactoryRegistry`. After realizing that I decided to put that into code, so now the `Map` is an `ImmutableMap` and created during construction of `CategoryRegistry`.